lib/assert.js

```diff
@@ -314,21 +320,29 @@ function setEquiv(a, b, strict, memo) {
   if (a.size !== b.size)
     return false;
 
+  // This is a set of the entries in b which have been consumed in our pairwise
+  // comparison. Initialized lazily so sets which only have value types can
+  // skip an extra allocation.
+  let bEntriesUsed = null;
+
   for (const val1 of a) {
     // If the value doesn't exist in the second set by reference, and its an
     // object or an array we'll need to go hunting for something thats
     // deep-equal to it. Note that this is O(n^2) complexity, and will get
     // slower if large, very similar sets / maps are nested inside.
     // Unfortunately there's no real way around this.
-    if (!setHasSimilarElement(b, val1, strict, memo)) {
+    if (bEntriesUsed == null && typeof val1 === 'object')
```

Comment:
The bug doesn't affect value types because its impossible for two strings, numbers, bools, etc to be deepEqual to one another without being reference-equal. So in the case where the set or map only contains value typed entries, there's no need to allocate an extra set for `bEntriesUsed`.

Lazily allocating the set when needed is a microoptimization.